### ISSUE TITLE
go/types, x/tools/go/ssa: panic: type param without replacement encountered

### ISSUE THREAD

**Issue by josharian** [golang/go#63260]
Run `ssa.(*Program).Build` on this module (for us, via govulncheck):

`go.mod`, does not reproduce with go 1.20:

```
module z

go 1.21
```

`main.go`, simplified from real code:

```
package main

func main() {
	use(f[*string])
}

func use(fn func() any) {}

func f[I interface{ *T }, T any]() any {
	var z T
	return z
}
```

Result:

`panic: type param without replacement encountered`

Reproduced with x/tools v0.13.0 and master as of Sep 27, 2023.

cc @alandonovan

**Comment by findleyr** [comment]
Reproducer: 
https://go.dev/play/p/LI7HtjPdNqe

Running on 1.21 yields "T false", yet on 1.20 yields "T true". Per @timothy-king all that matters in the go.mod language version, so it may relate to new unification logic. Note that if you comment out the inferred use in main, you get "T true" again.

CC @griesemer -- this is a go/types bug.

**Comment by findleyr** [comment]
My guess: something in the new unification logic is mutating the generic type.

**Comment by griesemer** [comment]
Type inference does parameter renaming. It may record the renamed type parameters.

**Comment by griesemer** [comment]
@gopherbot please consider this for backport to 1.21.3. This is a type-checker bug.

**Comment by dmitshur** [comment]
Moving this to the Go 1.22 milestone to track the fix at tip, and copying @rfindley's release-blocker label to the backport issue (which now tracks the fix for Go 1.21.3). Please feel free to remove that label if it's no longer needed.
